packages/liveblocks-core/src/connection.ts

```diff
@@ -605,44 +653,39 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
         // Stop retrying if this promise explicitly tells us so. This should,
         // in the case of a WebSocket connection attempt only be the case if
         // there is a configuration error.
-        //
-        // If a WebSocket connection is refused by the server, it could be that
-        // the token is expired or revoked, etc. In those cases, always go back
-        // to the authentication endpoint to try to obtain a new token.
         if (err instanceof StopRetrying) {
           return {
             target: "@idle.failed",
             effect: log(LogLevel.ERROR, err.message),
           };
         }
 
+        // Alternatively, it can happen if the server refuses the connection
+        // after accepting-then-closing the connection.
+        if (isCloseEvent(err) && err.code === 4999 /* Close Without Retry */) {
+          return {
+            target: "@idle.failed",
+            effect: log(LogLevel.ERROR, err.reason),
+          };
+        }
+
+        // We may skip re-authenticating only if this is a close event with
+        // a known custom close code, *except* when it's a 4001 (explicitly Not
+        // Allowed).
+        if (isCustomCloseEvent(err) && err.code !== 4001 /* Not Allowed */) {
+          return {
+            target: "@connecting.backoff",
+            effect: [
+              increaseBackoffDelayAggressively,
+              logPrematureErrorOrCloseEvent(err),
+            ],
+          };
+        }
```

Comment:
Similarly, these two cases address bugs. If the connection is refused[^1] with a 4999, we stop the machine and disconnect permanently. If we get a 4001, we re-authenticate. In all other cases, we can just retry after a backoff delay.

[^1]: This is what I've started to call the pattern of the server accepting-then-immediately-closing the WS connection with a close code. This verb matches the function name we already use in the backend.